pandas/core/indexes/base.py

```diff
@@ -2254,25 +2276,36 @@ def union(self, other):
         >>> idx2 = pd.Index([3, 4, 5, 6])
         >>> idx1.union(idx2)
         Int64Index([1, 2, 3, 4, 5, 6], dtype='int64')
+
+        >>> idx1 = pd.Index(['a', 'b', 'c', 'd'])
+        >>> idx2 = pd.Index([1, 2, 3, 4])
+        >>> idx1.union(idx2)
+        Index(['a', 'b', 'c', 'd', 1, 2, 3, 4], dtype='object')
         """
         self._assert_can_do_setop(other)
+
+        if not self._is_compatible_with_other(other):
+            return self._union_incompatible_dtypes(other)
+
+        # This line needs to be after _union_incompatible_dtypes to ensure
+        # the original type of other is not lost after being cast to Index
         other = ensure_index(other)
+        return self._union(other)
 
-        if len(other) == 0 or self.equals(other):
+    def _union(self, other):
+        """
+        Specific union logic should go here. In subclasses union behavior
+        should be overwritten here rather than in `self.union`
+        """
+        self._assert_can_do_setop(other)
+        other = ensure_index(other)
```

Comment:
I think we guarantee that other is an Index here. so can remove the ensure_index(other) here (and in subclasses)

Reply:
removed.